File: src/main/java/algo/breadthFirstSearch/WordLadder_127.java
package algo.breadthFirstSearch;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;

public class WordLadder_127 {
  /**
   *  Same as SlidingPuzzle
   */
  public static int ladderLength(String beginWord, String endWord, List<String> wordList) {
    var words = new LinkedList<String>();
    var visited = new HashSet<String>();
    words.offer(beginWord);
    visited.add(beginWord);
    var num = 1;
    while (!words.isEmpty()) {
      var size = words.size();
      while (size-- > 0) {
        var word = words.poll();
        if (word.equals(endWord)) {
          return num;
        }
        for (var w : next(wordList, word)) {
          if (visited.add(w)) {
            words.offer(w);
          }
        }
      }
      num++;
    }
    return 0;
  }

  private static List<String> next(List<String> wordList, String word) {
    var result = new ArrayList<String>();
    for (var w : wordList) {
      var diff = 0;
      for (var i = 0; i < w.length(); i++) {
        if (w.charAt(i) != word.charAt(i)) {
          diff++;
        }
      }
      if (diff == 1) {
        result.add(w);
      }
    }
    return result;
  }

  public static void main(String[] args) {
    System.out.println(ladderLength("hit", "cog", List.of("hot","dot","dog","lot","log","cog")));
    System.out.println(ladderLength("hit", "cog", List.of("hot","dot","dog","lot","log")));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/WordLadder_127.kt
package algo.breadthFirstSearch

import java.util.LinkedList

class KtWordLadder_127 {
    fun ladderLength(beginWord: String, endWord: String, wordList: List<String>): Int {
        fun nextWord(word: String, wordList: List<String>): List<String> {
            val words: MutableList<String> = mutableListOf()
            wordList.forEach { w ->
                var diff = 0
                for (i in w.indices) {
                    if (word[i] != w[i]) {
                        diff++
                    }
                    if (diff > 1) {
                        break
                    }
                }
                if (diff == 1) {
                    words.add(w)
                }
            }
            return words
        }

        var num = 1
        val visited: MutableSet<String> = HashSet()
        val q = LinkedList<String>()
        q.offer(beginWord)
        while (q.isNotEmpty()) {
            val size = q.size
            for (i in 1..size) {
                val word = q.poll()
                if (word == endWord) {
                    return num
                }
                if (!visited.add(word)) {
                    continue
                }
                nextWord(word, wordList).forEach { q.offer(it) }
            }
            num++
        }
        return 0
    }
}

fun main() {
    val ladder = KtWordLadder_127()
    println(ladder.ladderLength("hit", "cog", listOf("hot","dot","dog","lot","log","cog")))
}